src/auth/src/credentials.rs

```diff
@@ -137,6 +139,108 @@ impl Credentials {
+    // will not do.
+    //
+    // They also need to derive `Clone`, as the
+    // `gax::http_client::ReqwestClient`s which hold them derive `Clone`. So a
+    // `Box` will not do.
+    inner: Arc<dyn dynamic::AccessTokenCredentialsProvider>,
+}
+
+impl<T> std::convert::From<T> for AccessTokenCredentials
+where
+    T: crate::credentials::AccessTokenCredentialsProvider + Send + Sync + 'static,
+{
+    fn from(value: T) -> Self {
+        Self {
+            inner: Arc::new(value),
+        }
+    }
+}
+
+impl AccessTokenCredentials {
+    pub async fn access_token(&self) -> Result<AccessToken> {
+        self.inner.access_token().await
+    }
+}
+
+/// Makes [AccessTokenCredentials] compatible with clients that expect
+/// a [Credentials] and/or a [CredentialsProvider].
+impl CredentialsProvider for AccessTokenCredentials {
+    async fn headers(&self, extensions: Extensions) -> Result<CacheableResource<HeaderMap>> {
+        self.inner.headers(extensions).await
+    }
+
+    async fn universe_domain(&self) -> Option<String> {
+        self.inner.universe_domain().await
+    }
+}
+
+/// Represents an OAuth 2.0 access token.
+#[derive(Clone)]
+pub struct AccessToken {
+    /// The access token string.
+    pub token: String,
+    /// The type of the access token.
+    pub(crate) token_type: String,
+    /// The time when the access token expires.
+    pub(crate) expires_at: Option<Instant>,
+}
+
+impl std::fmt::Debug for AccessToken {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("AccessToken")
+            .field("token", &"[censored]")
+            .field("token_type", &self.token_type)
+            .field("expires_at", &self.expires_at)
+            .finish()
+    }
+}
+
+impl std::convert::From<CacheableResource<Token>> for Result<AccessToken> {
+    fn from(token: CacheableResource<Token>) -> Self {
+        match token {
+            CacheableResource::New { data, .. } => Ok(data.into()),
+            CacheableResource::NotModified => Err(errors::CredentialsError::from_msg(
+                false,
+                "Expecting token to be present",
+            )),
+        }
+    }
+}
+
+impl std::convert::From<Token> for AccessToken {
+    fn from(token: Token) -> Self {
+        Self {
+            token: token.token,
+            token_type: token.token_type,
+            expires_at: token.expires_at,
+        }
+    }
+}
+
+/// A trait for credential types that can provide direct access to an access token.
+///
+/// This trait is primarily intended for interoperability with other libraries that
+/// require a raw access token, or for calling Google Cloud APIs that are not yet
+/// supported by the SDK.
+pub trait AccessTokenCredentialsProvider: CredentialsProvider + std::fmt::Debug {
+    /// Asynchronously retrieves an access token.
+    fn access_token(&self) -> impl Future<Output = Result<AccessToken>> + Send;
+}
+
```

Comment:
hmmm... I was questioning whether we need this `pub trait`.

We needed a public trait for `Credentials` was mainly because `gax` unit tests needed a way to mock `Credentials`.

I guess I buy that apps built on top of these `AccessTokenCredentials` need something similar for their unit tests.

If so, we should add a quick mocking test to appease the coverage bot. Something like: https://github.com/googleapis/google-cloud-rust/blob/6709447d1934c9dad742f065cb9bca8b46bf7194/src/auth/tests/credentials.rs#L441-L465

Reply:
feel free to defer to a follow up PR

Reply:
that was a good point and I found out that the new trait was not mock friendly, since `AccessToken` had some private fields, so I decided to remove them for now.